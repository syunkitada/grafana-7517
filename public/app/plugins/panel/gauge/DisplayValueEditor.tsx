// Libraries
import React, { PureComponent } from 'react';

// Components
import { FormField, FormLabel, PanelOptionsGroup, UnitPicker } from '@grafana/ui';

// Types
import { DisplayValueOptions } from '@grafana/ui/src/utils/valueProcessor';

const labelWidth = 6;

export interface Props {
  options: DisplayValueOptions;
  onChange: (options: DisplayValueOptions) => void;
}

export class DisplayValueEditor extends PureComponent<Props> {
  onUnitChange = unit => this.props.onChange({ ...this.props.options, unit: unit.value });

  onDecimalChange = event => {
    if (!isNaN(event.target.value)) {
      this.props.onChange({
        ...this.props.options,
        decimals: parseInt(event.target.value, 10),
      });
    } else {
      this.props.onChange({
        ...this.props.options,
        decimals: null,
      });
    }
  };

  onPrefixChange = event => this.props.onChange({ ...this.props.options, prefix: event.target.value });
  onSuffixChange = event => this.props.onChange({ ...this.props.options, suffix: event.target.value });

  render() {
    const { unit, decimals, prefix, suffix } = this.props.options;

    let decimalsString = '';
    if (Number.isFinite(decimals)) {
      decimalsString = decimals.toString();
    }

    return (
      <PanelOptionsGroup title="Display Value">
        <div className="gf-form">
          <FormLabel width={labelWidth}>Unit</FormLabel>
          <UnitPicker defaultValue={unit} onChange={this.onUnitChange} />
        </div>
        <FormField
          label="Decimals"
          labelWidth={labelWidth}
          placeholder="auto"
          onChange={this.onDecimalChange}
          value={decimalsString}
          type="number"
        />
        <FormField label="Prefix" labelWidth={labelWidth} onChange={this.onPrefixChange} value={prefix || ''} />
        <FormField label="Suffix" labelWidth={labelWidth} onChange={this.onSuffixChange} value={suffix || ''} />
      </PanelOptionsGroup>
    );
  }
}
