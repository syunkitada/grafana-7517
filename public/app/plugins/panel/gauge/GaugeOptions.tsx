import React, { PureComponent } from 'react';
import { Switch } from 'app/core/components/Switch/Switch';
import { OptionModuleProps } from './module';
import { Label } from '../../../core/components/Label/Label';

export default class GaugeOptions extends PureComponent<OptionModuleProps> {
  onToggleThresholdLabels = () =>
    this.props.onChange({ ...this.props.options, showThresholdLabels: !this.props.options.showThresholdLabels });

  onToggleThresholdMarkers = () =>
    this.props.onChange({ ...this.props.options, showThresholdMarkers: !this.props.options.showThresholdMarkers });

  onMinValueChange = ({ target }) => this.props.onChange({ ...this.props.options, minValue: target.value });

  onMaxValueChange = ({ target }) => this.props.onChange({ ...this.props.options, maxValue: target.value });

  render() {
    const { maxValue, minValue, showThresholdLabels, showThresholdMarkers } = this.props.options;

    return (
      <div className="section gf-form-group">
        <h5 className="page-heading">Gauge</h5>
        <div className="gf-form-inline">
          <Switch
            label="Threshold labels"
            labelClass="width-10"
            checked={showThresholdLabels}
            onChange={this.onToggleThresholdLabels}
          />
        </div>
        <div className="gf-form-inline">
          <Switch
            label="Threshold markers"
            labelClass="width-10"
            checked={showThresholdMarkers}
            onChange={this.onToggleThresholdMarkers}
          />
        </div>
        <div className="gf-form-inline">
          <Label width={6}>Min value</Label>
          <input type="text" className="gf-form-input width-12" onChange={this.onMinValueChange} value={minValue} />
        </div>
        <div className="gf-form-inline">
          <Label width={6}>Max value</Label>
          <input type="text" className="gf-form-input width-12" onChange={this.onMaxValueChange} value={maxValue} />
        </div>
      </div>
    );
  }
}
