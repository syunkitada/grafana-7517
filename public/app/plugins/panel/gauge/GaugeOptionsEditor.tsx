import React, { PureComponent } from 'react';
import {
  FormField,
  FormLabel,
  PanelOptionsProps,
  PanelOptionsGroup,
  Select,
  SelectOptionItem,
  Switch,
} from '@grafana/ui';

import { GaugeOptions } from './types';

export default class GaugeOptionsEditor extends PureComponent<PanelOptionsProps<GaugeOptions>> {
  directionOptions: SelectOptionItem[] = [
    { value: 'auto', label: 'Auto' },
    { value: 'horizontal', label: 'Horizontal' },
    { value: 'vertical', label: 'Vertical' },
  ];

  labelWidth = 9;

  onToggleThresholdLabels = () =>
    this.props.onChange({ ...this.props.options, showThresholdLabels: !this.props.options.showThresholdLabels });

  onToggleThresholdMarkers = () =>
    this.props.onChange({ ...this.props.options, showThresholdMarkers: !this.props.options.showThresholdMarkers });

  onMinValueChange = ({ target }) => this.props.onChange({ ...this.props.options, minValue: target.value });

  onMaxValueChange = ({ target }) => this.props.onChange({ ...this.props.options, maxValue: target.value });

  onDirectionChange = ({ value }) => this.props.onChange({ ...this.props.options, direction: value });

  render() {
    const { options } = this.props;
    const { direction, maxValue, minValue, showThresholdLabels, showThresholdMarkers } = options;

    return (
      <PanelOptionsGroup title="Gauge">
        <FormField label="Min value" labelWidth={this.labelWidth} onChange={this.onMinValueChange} value={minValue} />
        <FormField label="Max value" labelWidth={this.labelWidth} onChange={this.onMaxValueChange} value={maxValue} />
        <Switch
          label="Show labels"
          labelClass={`width-${this.labelWidth}`}
          checked={showThresholdLabels}
          onChange={this.onToggleThresholdLabels}
        />
        <Switch
          label="Show markers"
          labelClass={`width-${this.labelWidth}`}
          checked={showThresholdMarkers}
          onChange={this.onToggleThresholdMarkers}
        />
        <div className="gf-form">
          <FormLabel width={this.labelWidth}>Direction</FormLabel>
          <Select
            defaultValue={this.directionOptions[0]}
            onChange={this.onDirectionChange}
            options={this.directionOptions}
            value={this.directionOptions.find(option => option.value === direction)}
            width={12}
          />
        </div>
      </PanelOptionsGroup>
    );
  }
}
