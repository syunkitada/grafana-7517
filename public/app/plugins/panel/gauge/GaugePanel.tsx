// Libraries
import React, { PureComponent } from 'react';

// Services & Utils
import { config } from 'app/core/config';

// Components
import { Gauge, DataLinksContextMenu, VizRepeater, VizRepeaterRenderValueProps } from '@grafana/ui';

// Types
import { GaugeOptions } from './types';
import { FieldDisplay, getFieldDisplayValues, VizOrientation, PanelProps } from '@grafana/data';
import { getFieldLinksSupplier } from 'app/features/panel/panellinks/linkSuppliers';

export class GaugePanel extends PureComponent<PanelProps<GaugeOptions>> {
  renderValue = (valueProps: VizRepeaterRenderValueProps<FieldDisplay>): JSX.Element => {
    const { options } = this.props;
    const { value, width, height } = valueProps;
    const { field, display } = value;

    return (
      <DataLinksContextMenu links={getFieldLinksSupplier(value)}>
        {({ openMenu, targetClassName }) => {
          return (
            <Gauge
              value={display}
              width={width}
              height={height}
              field={field}
              showThresholdLabels={options.showThresholdLabels}
              showThresholdMarkers={options.showThresholdMarkers}
              theme={config.theme}
              onClick={openMenu}
              className={targetClassName}
            />
          );
        }}
      </DataLinksContextMenu>
    );
  };

  getValues = (): FieldDisplay[] => {
    const { data, options, replaceVariables, fieldConfig } = this.props;
    return getFieldDisplayValues({
      fieldConfig,
      reduceOptions: options.reduceOptions,
      replaceVariables,
      theme: config.theme,
      data: data.series,
      autoMinMax: true,
    });
  };

  render() {
    const { height, width, data, renderCounter } = this.props;
    return (
      <VizRepeater
        getValues={this.getValues}
        renderValue={this.renderValue}
        width={width}
        height={height}
        source={data}
        renderCounter={renderCounter}
        orientation={VizOrientation.Auto}
      />
    );
  }
}
