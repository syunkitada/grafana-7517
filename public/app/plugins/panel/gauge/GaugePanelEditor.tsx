// Libraries
import React, { PureComponent } from 'react';
import {
  PanelOptionsGrid,
  FieldDisplayEditor,
  Switch,
  PanelOptionsGroup,
  FieldPropertiesEditor,
  ThresholdsEditor,
  LegacyValueMappingsEditor,
  DataLinksEditor,
} from '@grafana/ui';
import {
  PanelEditorProps,
  ReduceDataOptions,
  ThresholdsConfig,
  DataLink,
  FieldConfig,
  ValueMapping,
} from '@grafana/data';

import { GaugeOptions } from './types';
import {
  getCalculationValueDataLinksVariableSuggestions,
  getDataLinksVariableSuggestions,
} from '../../../features/panel/panellinks/link_srv';
import { NewPanelEditorContext } from '../../../features/dashboard/components/PanelEditor/PanelEditor';

export class GaugePanelEditor extends PureComponent<PanelEditorProps<GaugeOptions>> {
  labelWidth = 6;

  onToggleThresholdLabels = () =>
    this.props.onOptionsChange({ ...this.props.options, showThresholdLabels: !this.props.options.showThresholdLabels });

  onToggleThresholdMarkers = () =>
    this.props.onOptionsChange({
      ...this.props.options,
      showThresholdMarkers: !this.props.options.showThresholdMarkers,
    });

  onDisplayOptionsChanged = (
    fieldOptions: ReduceDataOptions,
    event?: React.SyntheticEvent<HTMLElement>,
    callback?: () => void
  ) => {
    this.props.onOptionsChange(
      {
        ...this.props.options,
        reduceOptions: fieldOptions,
      },
      callback
    );
  };

  onThresholdsChanged = (thresholds: ThresholdsConfig) => {
    const current = this.props.fieldConfig;
    this.props.onFieldConfigChange({
      ...current,
      defaults: {
        ...current.defaults,
        thresholds,
      },
    });
  };

  onValueMappingsChanged = (mappings: ValueMapping[]) => {
    const current = this.props.fieldConfig;
    this.props.onFieldConfigChange({
      ...current,
      defaults: {
        ...current.defaults,
        mappings,
      },
    });
  };

  onDataLinksChanged = (links: DataLink[]) => {
    const current = this.props.fieldConfig;
    this.props.onFieldConfigChange({
      ...current,
      defaults: {
        ...current.defaults,
        links,
      },
    });
  };

  onDefaultsChange = (field: FieldConfig) => {
    this.props.onFieldConfigChange({
      ...this.props.fieldConfig,
      defaults: field,
    });
  };

  render() {
    const { options, fieldConfig } = this.props;
    const { showThresholdLabels, showThresholdMarkers, reduceOptions: valueOptions } = options;

    const { defaults } = fieldConfig;

    const suggestions = valueOptions.values
      ? getDataLinksVariableSuggestions(this.props.data.series)
      : getCalculationValueDataLinksVariableSuggestions(this.props.data.series);

    return (
      <NewPanelEditorContext.Consumer>
        {useNewEditor => {
          if (useNewEditor) {
            return null;
          }

          return (
            <>
              <PanelOptionsGrid>
                <PanelOptionsGroup title="Display">
                  <FieldDisplayEditor
                    onChange={this.onDisplayOptionsChanged}
                    value={valueOptions}
                    labelWidth={this.labelWidth}
                  />
                  <Switch
                    label="Labels"
                    labelClass={`width-${this.labelWidth}`}
                    checked={showThresholdLabels}
                    onChange={this.onToggleThresholdLabels}
                  />
                  <Switch
                    label="Markers"
                    labelClass={`width-${this.labelWidth}`}
                    checked={showThresholdMarkers}
                    onChange={this.onToggleThresholdMarkers}
                  />
                </PanelOptionsGroup>

                <PanelOptionsGroup title="Field">
                  <FieldPropertiesEditor
                    showMinMax={true}
                    showTitle={true}
                    onChange={this.onDefaultsChange}
                    value={defaults}
                  />
                </PanelOptionsGroup>

                <ThresholdsEditor onChange={this.onThresholdsChanged} thresholds={defaults.thresholds} />
              </PanelOptionsGrid>

              <LegacyValueMappingsEditor onChange={this.onValueMappingsChanged} valueMappings={defaults.mappings} />
              <PanelOptionsGroup title="Data links">
                <DataLinksEditor
                  value={defaults.links}
                  onChange={this.onDataLinksChanged}
                  suggestions={suggestions}
                  maxLinks={10}
                />
              </PanelOptionsGroup>
            </>
          );
        }}
      </NewPanelEditorContext.Consumer>
    );
  }
}
