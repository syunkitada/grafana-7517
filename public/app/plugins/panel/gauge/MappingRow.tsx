import React, { PureComponent } from 'react';
import { Label } from 'app/core/components/Label/Label';
import { Select } from '@grafana/ui';
import { MappingType, RangeMap, ValueMap } from 'app/types';

interface Props {
  mapping: ValueMap | RangeMap;
  updateMapping: (mapping) => void;
  removeMapping: () => void;
}

interface State {
  from: string;
  id: number;
  operator: string;
  text: string;
  to: string;
  type: MappingType;
  value: string;
}

const mappingOptions = [
  { value: MappingType.ValueToText, label: 'Value' },
  { value: MappingType.RangeToText, label: 'Range' },
];

export default class MappingRow extends PureComponent<Props, State> {
  constructor(props) {
    super(props);

    this.state = {
      ...props.mapping,
    };
  }

  onMappingValueChange = event => {
    this.setState({ value: event.target.value });
  };

  onMappingFromChange = event => {
    this.setState({ from: event.target.value });
  };

  onMappingToChange = event => {
    this.setState({ to: event.target.value });
  };

  onMappingTextChange = event => {
    this.setState({ text: event.target.value });
  };

  onMappingTypeChange = mappingType => {
    this.setState({ type: mappingType });
  };

  updateMapping = () => {
    this.props.updateMapping({ ...this.state });
  };

  renderRow() {
    const { from, text, to, type, value } = this.state;

    if (type === MappingType.RangeToText) {
      return (
        <>
          <div className="gf-form">
            <Label width={4}>From</Label>
            <input
              className="gf-form-input width-8"
              value={from}
              onBlur={this.updateMapping}
              onChange={this.onMappingFromChange}
            />
          </div>
          <div className="gf-form">
            <Label width={4}>To</Label>
            <input
              className="gf-form-input width-8"
              value={to}
              onBlur={this.updateMapping}
              onChange={this.onMappingToChange}
            />
          </div>
          <div className="gf-form">
            <Label width={4}>Text</Label>
            <input
              className="gf-form-input width-10"
              value={text}
              onBlur={this.updateMapping}
              onChange={this.onMappingTextChange}
            />
          </div>
        </>
      );
    }

    return (
      <>
        <div className="gf-form">
          <Label width={4}>Value</Label>
          <input
            className="gf-form-input width-8"
            onBlur={this.updateMapping}
            onChange={this.onMappingValueChange}
            value={value}
          />
        </div>
        <div className="gf-form gf-form--grow">
          <Label width={4}>Text</Label>
          <input
            className="gf-form-input"
            onBlur={this.updateMapping}
            value={text}
            onChange={this.onMappingTextChange}
          />
        </div>
      </>
    );
  }

  render() {
    const { type } = this.state;

    return (
      <div className="gf-form-inline">
        <div className="gf-form">
          <Label width={5}>Type</Label>
          <Select
            placeholder="Choose type"
            isSearchable={false}
            options={mappingOptions}
            value={mappingOptions.find(o => o.value === type)}
            onChange={type => this.onMappingTypeChange(type.value)}
            width={7}
          />
        </div>
        {this.renderRow()}
        <div className="gf-form">
          <button onClick={this.props.removeMapping} className="gf-form-label gf-form-label--btn">
            <i className="fa fa-times" />
          </button>
        </div>
      </div>
    );
  }
}
