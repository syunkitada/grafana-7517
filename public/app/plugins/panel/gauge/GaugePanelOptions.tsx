// Libraries
import React, { PureComponent } from 'react';

// Components
import { ValueOptions } from 'app/plugins/panel/gauge/ValueOptions';
import { GaugeOptionsEditor } from './GaugeOptionsEditor';
import { ThresholdsEditor, ValueMappingsEditor } from '@grafana/ui';

// Types
import { PanelOptionsProps, Threshold, PanelOptionsGrid, ValueMapping } from '@grafana/ui';
import { GaugeOptions } from './types';

export class GaugePanelOptions extends PureComponent<PanelOptionsProps<GaugeOptions>> {
  onThresholdsChanged = (thresholds: Threshold[]) =>
    this.props.onChange({
      ...this.props.options,
      thresholds,
    });

  onValueMappingsChanged = (valueMappings: ValueMapping[]) =>
    this.props.onChange({
      ...this.props.options,
      valueMappings,
    });

  render() {
    const { onChange, options } = this.props;

    return (
      <>
        <PanelOptionsGrid>
          <ValueOptions onChange={onChange} options={options} />
          <GaugeOptionsEditor onChange={onChange} options={options} />
          <ThresholdsEditor onChange={this.onThresholdsChanged} thresholds={options.thresholds} />
        </PanelOptionsGrid>

        <ValueMappingsEditor onChange={this.onValueMappingsChanged} valueMappings={options.valueMappings} />
      </>
    );
  }
}
