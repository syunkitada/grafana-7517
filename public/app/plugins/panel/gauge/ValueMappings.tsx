import React, { PureComponent } from 'react';
import MappingRow from './MappingRow';
import { OptionModuleProps } from './module';
import { RangeMap, ValueMap } from 'app/types';

interface State {
  combinedMappings: any[];
  valueMaps: ValueMap[];
  rangeMaps: RangeMap[];
}

export default class ValueMappings extends PureComponent<OptionModuleProps, State> {
  constructor(props) {
    super(props);

    this.state = {
      combinedMappings: props.options.valueMaps.concat(props.options.rangeMaps),
      rangeMaps: props.options.rangeMaps,
      valueMaps: props.options.valueMaps,
    };
  }

  addMapping = () =>
    this.setState(prevState => ({
      combinedMappings: [...prevState.combinedMappings, { op: '', value: '', text: '' }],
    }));

  updateGauge = mapping => {
    this.setState(prevState => ({
      combinedMappings: prevState.combinedMappings.map(m => {
        if (m === mapping) {
          return { ...mapping };
        }

        return m;
      }),
    }));
  };

  render() {
    const { combinedMappings } = this.state;

    return (
      <div className="section gf-form-group">
        <h5 className="page-heading">Value mappings</h5>
        <div>
          {combinedMappings.length > 0 &&
            combinedMappings.map((mapping, index) => {
              return <MappingRow key={index} mapping={mapping} updateMapping={this.updateGauge} />;
            })}
        </div>
        <div className="add-mapping-row" onClick={this.addMapping}>
          <div className="add-mapping-row-icon">
            <i className="fa fa-plus" />
          </div>
          <div className="add-mapping-row-label">Add mapping</div>
        </div>
      </div>
    );
  }
}
