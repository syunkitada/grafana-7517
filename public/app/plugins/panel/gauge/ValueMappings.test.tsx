import React from 'react';
import { shallow } from 'enzyme';
import { GaugeOptions, MappingType, PanelOptionsProps } from '@grafana/ui';
import { defaultProps } from 'app/plugins/panel/gauge/GaugePanelOptions';

import ValueMappings from './ValueMappings';

const setup = (propOverrides?: object) => {
  const props: PanelOptionsProps<GaugeOptions> = {
    onChange: jest.fn(),
    options: {
      ...defaultProps.options,
      mappings: [
        { id: 1, operator: '', type: MappingType.ValueToText, value: '20', text: 'Ok' },
        { id: 2, operator: '', type: MappingType.RangeToText, from: '21', to: '30', text: 'Meh' },
      ],
    },
  };

  Object.assign(props, propOverrides);

  const wrapper = shallow(<ValueMappings {...props} />);

  const instance = wrapper.instance() as ValueMappings;

  return {
    instance,
    wrapper,
  };
};

describe('Render', () => {
  it('should render component', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('On remove mapping', () => {
  it('Should remove mapping with id 0', () => {
    const { instance } = setup();
    instance.onRemoveMapping(1);

    expect(instance.state.mappings).toEqual([
      { id: 2, operator: '', type: MappingType.RangeToText, from: '21', to: '30', text: 'Meh' },
    ]);
  });

  it('should remove mapping with id 1', () => {
    const { instance } = setup();
    instance.onRemoveMapping(2);

    expect(instance.state.mappings).toEqual([
      { id: 1, operator: '', type: MappingType.ValueToText, value: '20', text: 'Ok' },
    ]);
  });
});

describe('Next id to add', () => {
  it('should be 4', () => {
    const { instance } = setup();

    instance.addMapping();

    expect(instance.state.nextIdToAdd).toEqual(4);
  });

  it('should default to 1', () => {
    const { instance } = setup({ options: { ...defaultProps.options } });

    expect(instance.state.nextIdToAdd).toEqual(1);
  });
});
