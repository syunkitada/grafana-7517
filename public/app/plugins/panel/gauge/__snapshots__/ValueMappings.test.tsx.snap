// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="section gf-form-group"
>
  <h5
    className="page-heading"
  >
    Value mappings
  </h5>
  <div>
    <MappingRow
      key="Ok-0"
      mapping={
        Object {
          "id": 1,
          "operator": "",
          "text": "Ok",
          "type": 1,
          "value": "20",
        }
      }
      removeMapping={[Function]}
      updateMapping={[Function]}
    />
    <MappingRow
      key="Meh-1"
      mapping={
        Object {
          "from": "21",
          "id": 2,
          "operator": "",
          "text": "Meh",
          "to": "30",
          "type": 2,
        }
      }
      removeMapping={[Function]}
      updateMapping={[Function]}
    />
  </div>
  <div
    className="add-mapping-row"
    onClick={[Function]}
  >
    <div
      className="add-mapping-row-icon"
    >
      <i
        className="fa fa-plus"
      />
    </div>
    <div
      className="add-mapping-row-label"
    >
      Add mapping
    </div>
  </div>
</div>
`;
